### PR TITLE
feat: Validate amount with signer key authorizations

### PR DESCRIPTION
**Pull request by hstove**
This PR implements some of the changes discussed in https://github.com/stacks-network/stacks-core/discussions/4424.

- Closes https://github.com/stacks-network/stacks-core/issues/4422
- Closes https://github.com/stacks-network/stacks-core/issues/4417

The primary change is the addition of two fields to signer key authorization messages, which are validated either via a signature or a previously stored authorization. These fields are `max-amount` and `auth-id`. `max-amount` reflects the maximum amount of STX that can be used **in that specific stacking action**. The `auth-id` field is used to prevent replay attacks, which would allow a stacker to go beyond `max-amount` over multiple transactions.

Note that `stack-extend` doesn't verify the `amount` field in the authorization, because there are no additional STX added. In that function, the primary purpose of the authorization is around the number of reward cycles being extended.

`stack-increase` is also changed. It now includes the `signer-key`, along with the signature and related fields. Along with the same signature / authorization verifications that other stacking functions have, this function also checks to ensure that the `signer-key` parameter matches the `signer-key` for each updated cycle. **This means that you cannot call `stack-increase` if you've called `stack-extend` with a different signer key**.

In `pox-4` there is also a new map, `used-signer-key-authorizations`, which tracks whether a `(signer-key, pox-addr, reward-cycle, period, topic, max-amount, auth-id)` tuple has been used.

### CODE REVIEW

File: stackslib/src/chainstate/stacks/boot/pox-4.clar [hunk 1]
```diff
@@ -231,13 +234,33 @@
         period: uint,
         ;; A string representing the function where this authorization is valid. Either
         ;; `stack-stx`, `stack-extend`, or `agg-commit`.
-        topic: (string-ascii 12),
+        topic: (string-ascii 14),
```
Comment (jferrant) [inline stackslib/src/chainstate/stacks/boot/pox-4.clar:237]: Just out of curousity...why does the topic string need to increase in length from 12 to 14?

**Review by kantai — APPROVED** [review]

File: stackslib/src/chainstate/stacks/boot/pox-4.clar (hunk 1 above)
Comment (hstove) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:237]: It was only to allow for the whole string `stack-increase`, but definitely not strictly necessary. But this line points out that I need to update some of the documentation in this file (like this line).

**Comment by hstove** [comment]
@kantai I just re-requested review after pushing some documentation improvements

File: stackslib/src/chainstate/stacks/boot/pox-4.clar (hunk 1 above)
Comment (jferrant) [reply stackslib/src/chainstate/stacks/boot/pox-4.clar:237]: Ah yeah this makes sense. With the comment updated, totally get it.

**Review by jferrant — DISMISSED** [review]
Nice! :D extensive testing is great.

**Review by kantai — DISMISSED** [review]

**Review by setzeus — APPROVED** [review]
This is some great work here @hstove, good job on the thorough tests in pox_4_tests. Couldn't see anything in pox-4 or related works that called my attention.

LGTM.

**Review by zone117x — COMMENTED** [review]
If it's not too much of a hassle, could you add signer key to the pox4 `increase` synthetic event? If that's a pain we can do in a later PR

**Review by kantai — APPROVED** [review]

**Comment by hstove** [comment]
@zone117x good call and I should have caught that earlier. Let's get this merged and I made a new ticket: https://github.com/stacks-network/stacks-core/issues/4490

**Review by jferrant — APPROVED** [review]

**Pull request merged**